### DISCUSSION TITLE
Adding Early Data Support

### DISCUSSION THREAD

**Discussion by ad-chaos** [curl/curl#14013] [Ideas]
In #13528 @LaRuim floated the idea of adding early data support to curl, we are those undergrad students working on that and here we are with some ideas for a design

Looking at various SSL backend calls to APIs that send `early_data`, _must_ happen before invoking any kind of `connect` API. So far we’ve looked at the connection filter pipeline and it doesn’t seem to have a concept of `before_connect` and `after_connect`.

Given that knowledge, one way to add the ability to send early data would be to lie to the user, and based on whether a `CURLOPT_EARLY_DATA` option is set `do_connect` will in that case not connect and `do_send` will send the application data as early data then finish the handshake to the server, and if early data was rejected we send again normally, all in a single blocking call. Are we going in the right direction with this?

Making this async, where `write_early_data` is a `write` style API and `connect` another, would be more tricky and would require more investigation into curl’s internal state machine for progressing transfers.

Exposing the underlying SSL backend's early data API to a connection filter could be done by adding a new `write_early_data` method to the `Curl_ssl` vtls interface.

What are your thoughts?

**Comment by icing** [comment]
> In #13528 @LaRuim floated the idea of adding early data support to curl, we are those undergrad students working on that and here we are with some ideas for a design
> 
> Looking at various SSL backend calls to APIs that send `early_data`, _must_ happen before invoking any kind of `connect` API. So far we’ve looked at the connection filter pipeline and it doesn’t seem to have a concept of `before_connect` and `after_connect`.

I believe you also need a SSL session to make early data work. So, without one, everything should work as it is now, right?

> Given that knowledge, one way to add the ability to send early data would be to lie to the user, and based on whether a `CURLOPT_EARLY_DATA` option is set `do_connect` will in that case not connect and `do_send` will send the application data as early data then finish the handshake to the server, and if early data was rejected we send again normally, all in a single blocking call. Are we going in the right direction with this?
> 

Yes, I believe so. Delaying the handshake until the first `do_send()` is called seems the best way to approach this. In a first implementation, you could make the handshake blocking, since this seems simplest to see it working. For a solution usable in production, an async handling would be needed. But it seems ok to think about that once the early data handshake works.

> Making this async, where `write_early_data` is a `write` style API and `connect` another, would be more tricky and would require more investigation into curl’s internal state machine for progressing transfers.

Yes, I would not go there. Also, a transfer may have more than one SSL filter involved (proxying) and then where would one write to? Using `do_send()` is better as it respects the connection filter chain.
 
> Exposing the underlying SSL backend's early data API to a connection filter could be done by adding a new `write_early_data` method to the `Curl_ssl` vtls interface.
> 

Or do that in the vlts send() callback. But an additional method would also be fine, in case you find it better suited.

**Comment by ad-chaos** [comment]
> I believe you also need a SSL session to make early data work.

Unless we're mistaken, atleast for openssl, there's already [session resumption](https://github.com/curl/curl/blob/master/lib/vtls/openssl.c#L3948) in curl. The naming is unfortunate, but `ssl_sessionid` seems to hold the entire `SSL_SESSION` object which, when using tls1.3, holds all the relevant parameters in it, like `max_early_data`, a `PSK`, etc.

> So, without one, everything should work as it is now, right?

without a valid resumeable SSL Session, it would not be possible to send early data.
